[Tooling][Wear] Add Prototype Builds for Wear app

AliSoftware opened the pull request:
### Description

- Updates `release-toolkit` to `11.x`
   - So we can benefit from updated versions of the Prototype Build PR comment (including fixing the QR code after Google discontinued their googlecharts API a while ago)
- Fix `build_bundle` to build the expected app.
   - Indeed in the previous PRs adding Fastfile support for the Wear app, we forgot to update the name of the gradle task being called by the `build_bundle` lane, so it was always calling `bundleVanillaRelease`—instead of calling `:WooCommerce:bundleVanillaRelease` or `:WooCommerce-Wear:bundleVanillaRelease` based on the `app:` parameter, which it now does in this PR
- Add Prototype Build for Wear app
   - Not only so that one can test it more easily
   - But also so that we have at least one task running `WooCommerce-Wear:assemble` as part of the CI for every PR, so that PR checks also ensure that the Wear app will continue building and not have compilation errors in PRs updating related code.

### Testing Steps

 - Ensure that CI ends up posting **two** comments on this PR about Prototype Builds, one for the Mobile app (expanded by default), and one for the Wear app (collapsed by default, as I assumed that one would be used less often on most PRs so didn't want to make it too verbose)
 - Validate that the QR code is back on those PR comments
 - Try and install both Prototype Builds (mobile but also especially wear) on a test device to validate it works

AliSoftware committed:
Update release-toolkit to 11.x

AliSoftware committed:
Fix `build_bundle` to build the expected app

We forgot to update the gradle tasks being called to include `:#{app}:` in front of those (`:WooCommerce:assemble` / `:WooCommerce-Wear:assemble` instead of just `assemble`) to tell it to just build said module.

AliSoftware committed:
Add Prototype Build for Wear app

dangermattic commented on fastlane/Fastfile:
```diff
@@ -1109,47 +1109,49 @@ platform :android do
-  #####################################################################################
+  #
+  # @param [String] app The Android app to build (MOBILE_APP aka 'WooCommerce' or WEAR_APP aka 'WooCommerce-Wear')
+  # 
```
<table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/woocommerce/woocommerce-android/blob/91d1c1dea7e4edb98eb24c60779761d3ee69e4de/fastlane/Fastfile#L1116"></span>Layout/TrailingWhitespace: Trailing whitespace detected.</td>
    </tr>
  </tbody>
</table>

AliSoftware committed:
Fix rubocop violation

AliSoftware committed:
Fix call to `prototype-build.sh` in pipeline

 - `chmod +x` all the `.buildkite/commands/*` scripts (especially the `prototype-build.sh` one which didn't have the executable bit)
 - Fix the usage of quotes in the `command:` field of the pipeline

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  0 Messages
  1 Markdown
-->

## Danger has errored
[!] Invalid `Dangerfile` file: git `--git-dir=/var/lib/buildkite-agent/builds/ci-cluster-linter-i-074da126ca1a2fe72-1/automattic/woocommerce-android/.git` `--work-tree=/var/lib/buildkite-agent/builds/ci-cluster-linter-i-074da126ca1a2fe72-1/automattic/woocommerce-android` `-c` `core.quotePath=true` `-c` `color.ui=false` `cat-file` `-t` ``  2\>&1
status: pid 1166796 exit 128
output: "fatal: Not a valid object name \n"
```
 #  from Dangerfile:37
 #  -------------------------------------------
 #  
 >  pr_size_checker.check_diff_size(
 #    max_size: 300,
 #  -------------------------------------------
```

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

AliSoftware commented on fastlane/Fastfile:
```diff
@@ -1109,47 +1109,49 @@ platform :android do
     prototype_build_type = 'Debug'
 
     gradle(
-      task: 'assemble',
+      task: ":#{app}:assemble",
       flavor: prototype_flavor,
       build_type: prototype_build_type
     )
```
@ThomazFB @toupper @wzieba [From the error I get in CI](https://buildkite.com/automattic/woocommerce-android/builds/21235#018fce88-186a-42eb-a82f-89d93262172f/233-285), it seems that the Wear app does not have a `JalapenoDebug` variant to build the Prototype Build?

```
Cannot locate tasks that match ':WooCommerce-Wear:assembleJalapenoDebug' as task 'assembleJalapenoDebug' not found in project ':WooCommerce-Wear'.
```

Is this something you'd be able to add? (as you're way more knowledgeable in Gradle and Android project setup than I am 😅 ). Feel free to add commits to my branch to do the necessary changes in the gradle configs.

AliSoftware replied:
PS: My main goal for adding Prototype Build to the Wear app too was not only because I felt it would be useful for PR reviews if people wanted to test it… but also so that we had at least one CI job that compiled the `WooCommerce-Wear` module to ensure it still compiles at all time and didn't get compilation errors.

In case you end up deciding we don't want a Prototype Build for the Wear app after all (though why wouldn't we?), we'd still need to ensure that at least one of the CI jobs ends up running `assemble` on the the `WooCommerce-Wear` module (to ensure it always compiles on every PR); and we'd still need to land [this change](https://github.com/woocommerce/woocommerce-android/pull/11636#discussion_r1622364768) from this PR anyway (maybe I should have made two separate PRs stacked on top of one another for that… feel free to split that PR in two if that bit needs to land before we sort out the Prototype build bit of that PR)

ThomazFB replied:
Sure thing @AliSoftware! I just added the Jalapeno variant following the same configs from the main WooCommerce module. It's available directly through https://github.com/woocommerce/woocommerce-android/pull/11557. You can pull the changes from there and it should fix it.

AliSoftware replied:
Thanks! Merged the `feature/introduce-wear-app` branch into this PR and seems it indeed solved the issue and [allowed the Wear Prototype build to be generated](https://github.com/woocommerce/woocommerce-android/pull/11636#issuecomment-2142206688) 🎉

AliSoftware committed:
Fix flavor option handling in `build_bundle`

Some parts of that lane's implementation (especially `output_aab_file` variable) had `vanilla` flavor hardcoded instead of reading it from the `options[:flavor]` like the rest of the lane's implementation.

Also made that `flavor` option default to `Vanilla` if none was passed explicitly.

AliSoftware commented on fastlane/Fastfile:
```diff
@@ -843,30 +843,31 @@ platform :android do
-      UI.message("Building #{version} / #{build_code_current} - #{output_aab_file}...")
-      gradle(task: 'bundle', flavor: options[:flavor], build_type: 'Release')
+      gradle(task: 'lint', flavor: flavor, build_type: 'Release')
+      UI.message("Building #{app} #{version} / #{build_code_current} - #{output_aab_file}...")
+      gradle(task: ":#{app}:bundle", flavor: flavor, build_type: 'Release')
```
This is a change I missed in the previous PRs when updating the `build_bundle` lane to support Wear.

Which means that, despite calling `build_bundle` with an `app:` parameter, and using that `app` parameter to infer the name of the file generated by the gradle build… in practice until now we were still always calling the task just named `bundle` instead of `:#{app}:bundle` (i.e. `:WooCommerce:bundle` and `:WooCommerce-Wear:bundle`). As a result, iiuc, gradle were building the two apps (at least I assume that's what Gradle does in such cases) for each CI job (Mobile & Wear)… even if it then only used one of the two output files that had been generated (the one based on `output_dir` + `output_aab_file`, which are computed based on `app` parameter). 

So… quite a waste of build time for nothing 😅

AliSoftware committed:
More useful metadata in Prototype PR Comment

To differentiate Mobile vs Wear prototype builds/comments

AliSoftware commented on .buildkite/pipeline.yml:
💡 💭  Note for reviewers: you might want to review the diff on this particular file with "Ignore Whitespace" option turned on (via the ⚙️ menu at the top of the PR)

Indeed, most of the changes in this file was adding `- group` sections to group some steps together and thus increasing the indentation of the existing steps.

The only change that is not just an indentation change is the addition of the `Prototype Build: Wear App` step.

AliSoftware committed:
Merge branch 'feature/introduce-wear-app' into tooling/wear-app-prototype-build

AliSoftware commented:
Prototype Builds generation seems to have worked as expected on last commit 2731d0d 🎉 

[Prototype Build for Mobile app: same as usual (but with QR code fixed)](https://github.com/woocommerce/woocommerce-android/pull/11636#issuecomment-2141939933):
<img width="916" alt="image" src="https://github.com/woocommerce/woocommerce-android/assets/216089/eda521c1-5470-45b4-a22a-f960ff001ee7">

[Prototype Build for Wear app: comment](https://github.com/woocommerce/woocommerce-android/pull/11636#issuecomment-2142206688) collapsed by default, but can be expanded to reveal the similar info
<img width="924" alt="image" src="https://github.com/woocommerce/woocommerce-android/assets/216089/f8b62fc3-f9ef-49c9-aa70-afcad6fdf483">
<img width="928" alt="image" src="https://github.com/woocommerce/woocommerce-android/assets/216089/50a30366-0bf9-4ecf-862d-f3e0026e2b8f">

AliSoftware commented:
@ThomazFB can I let you test the Prototype Build for the Wear OS app by installing it on a test device (or emulator) to verify it works too? 🙇 

If so I think we'll be ready to have someone approve the PR (😇 ) so we can land it (disregarding the crash from Dangermattic which is odd but unrelated (see p1717157465536019-slack-C02KLTL3MKM)

ThomazFB commented:
Testing it right now, I'll be right back @AliSoftware.

ThomazFB commented:
I can confirm it's working as expected, @AliSoftware. Thank you for taking care of these changes so far 😄 

https://github.com/woocommerce/woocommerce-android/assets/5920403/2a6c3cf1-6a0a-4f99-bf4e-4db156a29fb7

ThomazFB commented:
@AliSoftware I think the addition of the Jalapeno variant introduced an error to the unit test pipeline. I'm unsure why, but I can see this happening here and in the main Watch app branch. Do you have an idea of why the prototype build could be failing there?

AliSoftware commented:
@ThomazFB This seems strange indeed as I see no immediate relationship at first at least.

Seems the failure that started to appear in the Unit Tests job is not a unit test failing per se but instead an error in the Code Coverage step (`jacoco`)?

Cc @wzieba in case you have some idea (you're probably more knowledgeable in JaCoCo than I am 😜)

wzieba committed:
Support JaCoCo in WooCommerce-Wear

wzieba commented on config/gradle/jacoco.gradle:
```diff
@@ -56,6 +60,19 @@ rootProject.afterEvaluate {
                 '**/*Screen*',
                 '**/*Screen$*',
+
+                // Classes below have the same name in WooCommerce and WooCommerce-Wear modules
+                // Without exclusions, JaCoCo fails with "Can't add different class with same name"
+                "**/AppCoroutineScope*",
+                "**/*DateUtils*",
+                "**/*NumberExt*",
+                "**/*DateExt*",
+                "**/*OrderEntityExt*",
+                "**/*CalendarExt*",
+
+                "**/*DataStoreType*",
+                "**/*DataStoreQualifier*",
+                "**/*AnalyticsTracker*",
```
<img width="1182" alt="Screenshot 2024-06-03 at 10 41 00" src="https://github.com/woocommerce/woocommerce-android/assets/5845095/c7998a54-ca78-48e2-b611-47ec1929bd89">

We can address it by e.g. introducing `com.woocommerce.android.wear` package and put all `wear` code there (except `WooCommerceWear` class, as I understand it _has_ to be in `com.woocommerce.android.app`).

Or we can introduce a new module, like `utils` or `ext` or anything else, that will share code between `WooCommerce` and `WooCommerce-Wear` modules. 

But for now, I'm proposing this fix to not block further progress.

AliSoftware replied:
> But for now, I'm proposing this fix to not block further progress.

I'm totally fine for us to implement one of the better/long-term approaches you suggest in a separate and subsequent PR so as not to block this PR from landing 👍 

That being said, that means that someone will need to implement such subsequent PR (to reorganize modules a bit and solve `jacoco` issue) before we land the main `feature/introduce-wear-app` branch into `trunk` and we launch the Wear app, otherwise any PR based on `trunk` would start failing on the Unit Tests's job and `jacoco` and will be blocked from being merged.

Can one of you (@wzieba , @ThomazFB ) thus take ownership of that and put in your plan to work on that subsequent PR before the Wear OS app launch? Just to be sure that (1) it's not forgotten and (2) that someone takes DRI role on this (instead of everyone thinking someone else will do it and nobody ending up doing it after all 😄 )

Thanks!

wzieba replied:
> That being said, that means that someone will need to implement such subsequent PR (to reorganize modules a bit and solve jacoco issue) before we land the main feature/introduce-wear-app branch into trunk and we launch the Wear app, otherwise any PR based on trunk would start failing on the Unit Tests's job and jacoco and will be blocked from being merged.

I'm not sure if I understood - could you please share more on why PR based on `trunk` would start failing the unit tests job? I think we can safely land `tooling/wear-app-prototype-build` into ` feature/introduce-wear-app` and then `feature/introduce-wear-app` to `trunk` and nothing should fail. Then we can think of how we can address this minor issue (no code coverage data for a few utils classes). WDYT?

AliSoftware replied:
Ohhh sorry I misread your comment above. I thought you were saying that we could keep the failure on `jacoco` that we had on this PR and land it as is and only fix the `jacoco` error (that led to the Unit Tests job failure on CI) in a subsequent PR… which is why I said that we'd still need to fix `jacoco` making the Unit Tests CI join fail before we land everything in `trunk`.

But that's because I missed that you already fixed that part and the `jacoco` failing the job on CI via https://github.com/woocommerce/woocommerce-android/pull/11636/commits/14cc55b978c98bcc48d36dcdebea9f011595760b after all, and your comment above  about "We can address this" was not referring to the CI job going red because of `jacoco`, but instead referring to a different failure ("Can't add different class  with same name"), which doesn't lead to the CI job going red anymore (but only means we'll be missing some code coverage data for a few classes), which is indeed way more minor than the whole CI job going red and preventing the PR to fail.

I think that this PR still being shown as red in GitHub (but for yet a different CI job, this time it's only the Danger check that is unhappy, not Unit Tests anymore, which I failed to see as now being green) also played a role in me misreading it all and thinking we still had the `jacoco` / Unit Tests CI job fail too, while we don't anymore 👍

wzieba replied:
ahh gotcha, I see the point you were making in previous comment now 👍 I'm glad we cleared this out! 🙌

AliSoftware replied:
Yeah tbh I should have given more attention to the fact that this comment was an inline comment precisely on the lines 64–75 of the `jacoco.gradle` config file that you made exactly to workaround the crash we'd have otherwise. 🤦 
While instead I was in catch-up mode from all the PR ping I had during my AFK and only saw it as a more "global" answer in the whole PR in reply to my "cc {you} in case you have some idea [to fix the jacoco CI failure]" [from above](https://github.com/woocommerce/woocommerce-android/pull/11636#issuecomment-2142630192)… I should have taken my ☕ and be more attentive what lines that inline comment was on 😅

AliSoftware commented on config/gradle/jacoco.gradle:
```diff
@@ -8,7 +8,11 @@ jacoco {
+        dependsOn(
+                ':WooCommerce:testJalapenoDebugUnitTest',
+                ':libs:cardreader:testDebugUnitTest',
+                ':WooCommerce-Wear:testJalapenoDebugUnitTest',
```
👍  Missed that fix, thanks for addressing it @wzieba !
